analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs

```diff
@@ -18,158 +18,147 @@
+    private static Location GetReportLocation(InvocationExpressionSyntax invocation, bool methodCalledAsStatic) =>
+        invocation.Expression switch
         {
-            if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+            MemberAccessExpressionSyntax memberAccess => methodCalledAsStatic
+                ? memberAccess.GetLocation()
+                : memberAccess.OperatorToken.CreateLocation(invocation),
+            _ => invocation.Expression.GetLocation()
+        };
+
+    private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol, SemanticModel semanticModel)
+    {
+        return CollectionExpression() is { } collection
+            ? semanticModel.GetTypeInfo(collection).Type switch
             {
-                return invocation.Expression.GetLocation();
+                INamedTypeSymbol { TypeArguments: { Length: 1 } typeArguments } => typeArguments.First(),
+                IArrayTypeSymbol { Rank: 1 } arrayType => arrayType.ElementType,    // casting is necessary for multidimensional arrays
+                _ => null
             }
+            : null;
 
-            return methodCalledAsStatic
-                ? memberAccess.GetLocation()
-                : memberAccess.OperatorToken.CreateLocation(invocation);
-        }
-
-        private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol,
-            SemanticModel semanticModel)
+        ExpressionSyntax CollectionExpression()
         {
-            ExpressionSyntax collection;
             if (methodSymbol.MethodKind == MethodKind.Ordinary)
             {
-                if (!invocation.ArgumentList.Arguments.Any())
+                if (invocation.ArgumentList.Arguments.FirstOrDefault() is { Expression: { } firstArgumentExpression })
                 {
-                    return null;
+                    return firstArgumentExpression;
                 }
-                collection = invocation.ArgumentList.Arguments.First().Expression;
             }
-            else
-            {
-                if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
-                {
-                    return null;
-                }
-                collection = memberAccess.Expression;
-            }
-
-            var typeInfo = semanticModel.GetTypeInfo(collection);
-            if (typeInfo.Type is INamedTypeSymbol collectionType &&
-                collectionType.TypeArguments.Length == 1)
-            {
-                return collectionType.TypeArguments.First();
-            }
-
-            if (typeInfo.Type is IArrayTypeSymbol arrayType &&
-                arrayType.Rank == 1) // casting is necessary for multidimensional arrays
+            else if (invocation is { Expression: MemberAccessExpressionSyntax { Expression: { } memberAccessExpression } })
```

Comment:
`GetParentConditionalAccessExpression` should help to fix some of the issues I described in the coverage comment above.